### ISSUE TITLE
关于Debiased Constraint代码

### ISSUE THREAD

**Issue by VerdureChen** [KID-22/Source-Bias#1]
您好，请问是否可公开Debiased Constraint的训练代码部分？

**Comment by KID-22** [comment]
以下是我们实验的一个版本的loss，其对sentence_transformers中的[MultipleNegativesRankingLoss](https://www.sbert.net/docs/package_reference/sentence_transformer/losses.html?highlight=loss#sentence_transformers.losses.MultipleNegativesRankingLoss)做了修改来进行debias:

```
class Debiased_MultipleNegativesRankingLoss(nn.Module):
    def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_fct = util.cos_sim, alpha: float = 0.1):
        """
        :param model: SentenceTransformer model
        :param scale: Output of similarity function is multiplied by scale value
        :param similarity_fct: similarity function between sentence embeddings. By default, cos_sim. Can also be set to dot product (and then set scale to 1)
        """
        super(Debiased_MultipleNegativesRankingLoss, self).__init__()
        self.model = model
        self.scale = scale
        self.similarity_fct = similarity_fct
        self.alpha = alpha
        self.cross_entropy_loss = nn.CrossEntropyLoss()
        self.margin_rank_loss = nn.MarginRankingLoss()

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        reps = [self.model(sentence_feature)['sentence_embedding'] for sentence_feature in sentence_features]
        query_rep, human_doc_rep, llm_doc_rep = reps
        
        human_scores = self.similarity_fct(query_rep, human_doc_rep) * self.scale
        llm_scores = self.similarity_fct(query_rep, llm_doc_rep) * self.scale
        human_pos_scores = torch.diag(human_scores).reshape(-1, 1)
        llm_pos_scores = torch.diag(llm_scores).reshape(-1, 1)

        mask = torch.ones_like(human_scores) - torch.eye(human_scores.shape[0], device=human_scores.device)
        human_neg_scores = human_scores[mask.bool()].reshape(-1, human_scores.shape[1]-1)
        llm_neg_scores = llm_scores[mask.bool()].reshape(-1, llm_scores.shape[1]-1)
        scores_1 = torch.cat([human_pos_scores, human_neg_scores, llm_neg_scores], dim=1)
        scores_2 = torch.cat([llm_pos_scores, human_neg_scores, llm_neg_scores], dim=1)
        labels_1 = torch.zeros(len(scores_1), dtype=torch.long, device=scores_1.device)
        labels_2 = torch.zeros(len(scores_2), dtype=torch.long, device=scores_1.device)
        loss = self.cross_entropy_loss(scores_1, labels_1) + self.cross_entropy_loss(scores_2, labels_2)
        
        constraint_loss = self.margin_rank_loss(human_pos_scores, llm_pos_scores, torch.ones_like(human_pos_scores))
        
        return loss + self.alpha * constraint_loss

    def get_config_dict(self):
        return {'scale': self.scale, 'similarity_fct': self.similarity_fct.__name__}
```

结合这个loss，你可以利用beir或者sentence_transformers的训练代码进行训练，只需要修改传入的loss即可，例如使用如下代码：
```
if args.score_func == "cos":
    train_loss = Debiased_MultipleNegativesRankingLoss(
        model=retriever.model, similarity_fct=util.cos_sim, alpha=alpha)
elif args.score_func == "dot":
    train_loss = Debiased_MultipleNegativesRankingLoss(
        model=retriever.model, similarity_fct=util.dot_score, alpha=alpha, scale=1)
```

详细代码已经上传，请见[this folder](https://github.com/KID-22/Source-Bias/tree/main/debias).
